bpf/cpu/cpu.bpf.c

```diff
@@ -58,24 +54,217 @@ typedef __u64 stack_trace_type[MAX_STACK_DEPTH];
+  __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+  __uint(max_entries, 1);
+  __type(key, __u32);
+  __type(value, stack_count_key_t);
+} heap SEC(".maps");
+
+struct {
+  __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+  __uint(max_entries, 10);
+  __type(key, __u32);
+  __type(value, __u32);
+} percpu_stats SEC(".maps");
 
 /*=========================== HELPER FUNCTIONS ==============================*/
 
+static void unwind_success() {
+  u32 *c = bpf_map_lookup_elem(&percpu_stats, &UNWIND_SUCCESS);
+  if (c != NULL) {
+    *c += 1;
+  }
+}
+
+static void unwind_truncated() {
+  u32 *c = bpf_map_lookup_elem(&percpu_stats, &UNWIND_TRUNCATED);
+  if (c != NULL) {
+    *c += 1;
+  }
+}
+
+static void unwind_unsupported_expression() {
+  u32 *c = bpf_map_lookup_elem(&percpu_stats, &UNWIND_UNSUPPORTED_EXPRESSION);
+  if (c != NULL) {
+    *c += 1;
+  }
+}
+
+static void unwind_should_never_happen_error() {
+  u32 *c =
+      bpf_map_lookup_elem(&percpu_stats, &UNWIND_SHOULD_NEVER_HAPPEN_ERROR);
+  if (c) {
+    *c += 1;
+  }
+}
+
+static void unwind_catchall_error() {
+  u32 *c = bpf_map_lookup_elem(&percpu_stats, &UNWIND_CATCHALL_ERROR);
+  if (c) {
+    *c += 1;
+  }
+}
+
+static void unwind_uncovered_error() {
+  u32 *c = bpf_map_lookup_elem(&percpu_stats, &UNWIND_PC_NOT_COVERED_ERROR);
+  if (c) {
+    *c += 1;
+  }
+}
+
+static void unwind_print_stats() {
+  u32 *success_counter = bpf_map_lookup_elem(&percpu_stats, &UNWIND_SUCCESS);
+  if (success_counter == NULL) {
+    return;
+  }
+
+  u32 *total_counter =
+      bpf_map_lookup_elem(&percpu_stats, &UNWIND_SAMPLES_COUNT);
+  if (total_counter == NULL) {
+    return;
+  }
+
+  u32 *truncated_counter =
+      bpf_map_lookup_elem(&percpu_stats, &UNWIND_TRUNCATED);
+  if (truncated_counter == NULL) {
+    return;
+  }
+
+  u32 *unsup_expression =
+      bpf_map_lookup_elem(&percpu_stats, &UNWIND_UNSUPPORTED_EXPRESSION);
+  if (unsup_expression == NULL) {
+    return;
+  }
+
+  u32 *not_covered_count =
+      bpf_map_lookup_elem(&percpu_stats, &UNWIND_PC_NOT_COVERED_ERROR);
+  if (not_covered_count == NULL) {
+    return;
+  }
+
+  u32 *catchall_count =
+      bpf_map_lookup_elem(&percpu_stats, &UNWIND_CATCHALL_ERROR);
+  if (catchall_count == NULL) {
+    return;
+  }
+
+  u32 *never =
+      bpf_map_lookup_elem(&percpu_stats, &UNWIND_SHOULD_NEVER_HAPPEN_ERROR);
+  if (never == NULL) {
+    return;
+  }
+
+  bpf_printk("[[ stats for cpu %d ]]", (int)bpf_get_smp_processor_id());
+  bpf_printk("success=%lu", *success_counter);
+  bpf_printk("unsup_expression=%lu", *unsup_expression);
+  bpf_printk("truncated_counter=%lu", *truncated_counter);
+  bpf_printk("catchall_count=%lu", *catchall_count);
+  bpf_printk("never=%lu", *never);
+
+  bpf_printk("total_counter=%lu", *total_counter);
+  bpf_printk("(not_covered_count=%lu)", *not_covered_count);
+}
+
+static void bump_samples() {
+  u32 *c = bpf_map_lookup_elem(&percpu_stats, &UNWIND_SAMPLES_COUNT);
+  if (c != NULL) {
+    *c += 1;
+    if (*c % 50 == 0) {
```

Comment:
Print statistics only every 50 stacks to reduce the output, it can be a bit overwhelming for processes that spend a lot of the time on-CPU